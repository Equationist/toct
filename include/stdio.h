/* stdio.h - Standard Input/Output Library for CompilerKit C Compiler */

#ifndef _STDIO_H
#define _STDIO_H

/* Standard I/O types */
typedef struct {
    int _fd;
    int _flags;
    char* _buffer;
} FILE;

/* Standard streams */
extern FILE* stdin;
extern FILE* stdout; 
extern FILE* stderr;

/* Standard I/O functions */
int printf(const char* format, ...);
int fprintf(FILE* stream, const char* format, ...);
int sprintf(char* str, const char* format, ...);
int snprintf(char* str, int size, const char* format, ...);

int scanf(const char* format, ...);
int fscanf(FILE* stream, const char* format, ...);
int sscanf(const char* str, const char* format, ...);

int putchar(int c);
int puts(const char* s);
int fputs(const char* s, FILE* stream);
int fputc(int c, FILE* stream);

int getchar(void);
char* gets(char* s);
char* fgets(char* s, int size, FILE* stream);
int fgetc(FILE* stream);
int ungetc(int c, FILE* stream);

FILE* fopen(const char* filename, const char* mode);
FILE* freopen(const char* filename, const char* mode, FILE* stream);
int fclose(FILE* stream);
int fflush(FILE* stream);

long ftell(FILE* stream);
int fseek(FILE* stream, long offset, int whence);
void rewind(FILE* stream);
int feof(FILE* stream);
int ferror(FILE* stream);
void clearerr(FILE* stream);

/* File position constants */
#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

/* End of file constant */
#define EOF (-1)

/* Buffer size */
#define BUFSIZ 8192

/* Null pointer */
#ifndef NULL
#define NULL ((void*)0)
#endif

#endif /* _STDIO_H */