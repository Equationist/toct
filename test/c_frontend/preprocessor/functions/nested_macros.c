/* Test nested macro expansion */
#define DOUBLE(x) ((x) * 2)
#define TRIPLE(x) ((x) * 3)
#define ADD(a, b) ((a) + (b))
#define MULTIPLY(a, b) ((a) * (b))

/* Nested macro definitions */
#define DOUBLE_TRIPLE(x) DOUBLE(TRIPLE(x))
#define ADD_DOUBLE(a, b) ADD(DOUBLE(a), DOUBLE(b))
#define QUAD(x) DOUBLE(DOUBLE(x))

#include <stdio.h>

int main() {
    int x = 5;
    
    printf("DOUBLE(%d) = %d\n", x, DOUBLE(x));
    printf("TRIPLE(%d) = %d\n", x, TRIPLE(x));
    printf("DOUBLE_TRIPLE(%d) = %d\n", x, DOUBLE_TRIPLE(x));
    printf("ADD_DOUBLE(%d, %d) = %d\n", x, 3, ADD_DOUBLE(x, 3));
    printf("QUAD(%d) = %d\n", x, QUAD(x));
    
    /* Complex nesting */
    printf("MULTIPLY(DOUBLE(%d), TRIPLE(%d)) = %d\n", x, x, MULTIPLY(DOUBLE(x), TRIPLE(x)));
    
    return 0;
}