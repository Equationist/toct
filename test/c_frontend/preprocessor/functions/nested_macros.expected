DOUBLE(5) = 10
TRIPLE(5) = 15
DOUBLE_TRIPLE(5) = 30
ADD_DOUBLE(5, 3) = 16
QUAD(5) = 20
MULTIPLY(DOUBLE(5), TRIPLE(5)) = 150